<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Inferring regulomes from multi-modal single-cell genomics data • Pando</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Inferring regulomes from multi-modal single-cell genomics data">
<meta property="og:description" content="Pando jointly uses scRNA-seq and scATAC-seq measurements to infer regulatory relationships between TF-binding site pairs and target genes. ">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">Pando</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/getting_started.html">Basic workflow</a>
    </li>
    <li>
      <a href="articles/visualization.html">Visualization</a>
    </li>
    <li>
      <a href="articles/regions.html">Selection of candidate regions</a>
    </li>
    <li>
      <a href="articles/models.html">Models for GRN inference</a>
    </li>
    <li>
      <a href="articles/association.html">Gene-peak association</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://bsse.ethz.ch/qdb" class="external-link">
    <span class="fas fa-university fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/quadbiolab/Pando" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="pando-">Pando <img src="reference/figures/logo.png" align="right" width="180"><a class="anchor" aria-label="anchor" href="#pando-"></a>
</h1></div>
<p>Pando leverages multi-modal singel-cell measurements to infer gene regulatory networks (GRNs) using a flexible modelling framework. By modeling the relationship between TF-binding site pairs with the expression of target genes, Pando simultaneously infers gene modules and sets of regulatory regions for each transcription factor.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'quadbiolab/Pando'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="about-pando">About Pando<a class="anchor" aria-label="anchor" href="#about-pando"></a>
</h2>
<p>The fate and state of a cell are regulated through complex circuits of transcription factors (TFs) converging at regulatory elements to enable precise control of gene expression. Modern single-cell genomic approaches allow the simultaneous profiling of gene expression and chromatin accessibility in individual cells, which opens up new opportunities for the inference of gene regulatory networks (GRNs).</p>
<p>The unifying idea behind many modern GRN inference methods is to model the expression of each gene as a function of TF abundances. The weights or coefficients of this model can then be interpreted as a measure of the regulatory interaction between TF and target gene. Additional (epi-) genomic information (such as predicted TF binding sites) is often used to constrain or refine the model.</p>
<p>Pando tries to generalize this concept to make use of the multi-modal nature of modern single-cell technologies by incorporating TF binding information directly into the model. By utilizing jointly measured or integrated scRNA-seq and scATAC-seq data, Pando models the expression of genes based on the interaction of TF expression with the accessibility of their putative binding site. By offering a number of different pre-processing and modeling choices, Pando strives to be a modular and flexible framework for single-cell GRN inference.</p>
</div>
<div class="section level2">
<h2 id="usage-overview">Usage overview<a class="anchor" aria-label="anchor" href="#usage-overview"></a>
</h2>
<div class="section level3">
<h3 id="initiating-the-grn">Initiating the GRN<a class="anchor" aria-label="anchor" href="#initiating-the-grn"></a>
</h3>
<p>Pando interacts directly with <a href="https://satijalab.org/seurat/" class="external-link">Seurat objects</a> and integrates well with <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a> and <a href="https://satijalab.org/signac/" class="external-link">Signac</a> workflows. To use Pando, you’ll need a Seurat object with two assays, one with scRNA-seq transcript counts and one with scATAC-seq peak accessibility. With this object (let’s call it <code>seurat_object</code>) ready, you can start off by initializing the GRN using the function <code><a href="reference/initiate_grn.html">initiate_grn()</a></code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/initiate_grn.html">initiate_grn</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span></span></code></pre></div>
<p>This will create a <code>RegulatoryNetwork</code> object inside the Seurat object and select candidate regulatory regions. By default, Pando will consider all peaks as putative regulatory regions, but the set of candidate regions can be constrained by providing a <code>GenomicRanges</code> object in the <code>regions</code> argument. Pando ships with a set of conserved regions (<code>phastConsElements20Mammals.UCSC.hg38</code>) as well as predicted regulatory elements from <a href="https://screen.encodeproject.org/" class="external-link">ENCODE</a> (<code>SCREEN.ccRE.UCSC.hg38</code>) for the human genome (hg38), which could be used here. However, one could also select candidate regions in other ways, for instance by using <a href="https://cole-trapnell-lab.github.io/cicero-release/" class="external-link">Cicero</a>.</p>
</div>
<div class="section level3">
<h3 id="finding-tf-binding-sites">Finding TF binding sites<a class="anchor" aria-label="anchor" href="#finding-tf-binding-sites"></a>
</h3>
<p>Once the <code>RegulatoryNetwork</code> object is initiated with candidate regions, we can scan for TF binding motifs in these regions by using the function <code><a href="reference/find_motifs.html">find_motifs()</a></code></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">motifs</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/find_motifs.html">find_motifs</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_object</span>,</span>
<span>    pfm <span class="op">=</span> <span class="va">motifs</span>,</span>
<span>    genome <span class="op">=</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This uses <a href="https://github.com/GreenleafLab/motifmatchr" class="external-link">motifmatchr</a> to pair up TFs with their putative binding sites. Pando provides a custom motif database (<code>motifs</code>) compiled from <a href="http://jaspar.genereg.net/" class="external-link">JASPAR</a> and <a href="http://cisbp.ccbr.utoronto.ca/" class="external-link">CIS-BP</a>, but in principle any <code>PFMatrixList</code> object can be provided here. A data frame with motif-to-TF assignments can be provided in the <code>motif_tfs</code> argument.</p>
</div>
<div class="section level3">
<h3 id="inferring-the-grn">Inferring the GRN<a class="anchor" aria-label="anchor" href="#inferring-the-grn"></a>
</h3>
<p>Now everything should be ready to infer the GRN by fitting regression models for the expression of each gene. In Pando, this can be done by using the function <code><a href="reference/infer_grn.html">infer_grn()</a></code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/infer_grn.html">infer_grn</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_object</span>,</span>
<span>    peak_to_gene_method <span class="op">=</span> <span class="st">'Signac'</span>,</span>
<span>    method <span class="op">=</span> <span class="st">'glm'</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Here, we first select regions near genes, either by simply considering a distance upstream and/or downstream of the gene (<code>peak_to_gene_method='Signac'</code>) or by also considering overlapping regulatory regions as is done by <a href="http://great.stanford.edu/public/html/" class="external-link">GREAT</a> (<code>peak_to_gene_method='GREAT'</code>).</p>
<p>You can also choose between a number of different models using the <code>method</code> argument, such as GLMs (<code>'glm'</code>) regularized GLMs <a href="https://glmnet.stanford.edu/articles/glmnet.html" class="external-link">(<code>'glmnet'</code>, <code>'cv.glmnet'</code>)</a> or Bayesian regression models <a href="https://paul-buerkner.github.io/brms/" class="external-link">(<code>'brms'</code>)</a>. We are also integrated gradient boosting regression with <a href="https://xgboost.readthedocs.io/en/latest/R-package/xgboostPresentation.html" class="external-link">XGBoost</a> as it is used by <a href="https://github.com/aertslab/GRNBoost" class="external-link">GRNBoost</a>/<a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> as well as bagging and bayesian ridge models with <a href="https://scikit-learn.org/stable/" class="external-link">scikit-learn</a> as they are used by <a href="https://github.com/morris-lab/CellOracle" class="external-link">CellOracle</a>.</p>
<p>Once the models are fit, model coefficients can be inspected with</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="module-extraction">Module extraction<a class="anchor" aria-label="anchor" href="#module-extraction"></a>
</h3>
<p>Based on the model coefficients, we can construct a network between TFs and target genes. This can be further summarized to construct gene and regulatory modules with the set of target genes and regulatory regions for each TF. In Pando we do this with</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/find_modules.html">find_modules</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span></span></code></pre></div>
<p>To access the extracted modules, you can use the function <code><a href="reference/NetworkModules.html">NetworkModules()</a></code>:</p>
<pre><code><span><span class="va">modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/NetworkModules.html">NetworkModules</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span></span>
<span><span class="va">modules</span><span class="op">@</span><span class="va">meta</span></span></code></pre>
<p>The <code>meta</code> slot holds a dataframe with module inforamtion.</p>
<p>If you are curious to find out more, check out our vignettes!</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/quadbiolab/Pando" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing Pando</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jonas Simon Fleck <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-4686-7254" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Jonas Simon Fleck.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
